server/util/canary/canary.go

```diff
@@ -0,0 +1,74 @@
+package canary
+
+import (
+	"flag"
+	"fmt"
+	"runtime"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+)
+
+var (
+	enableCanaries = flag.Bool("app.enable_canaries", true, "If true, enable slow function canaries")
+)
+
+type CancelFunc func()
+type CanaryFunc func(timeTaken time.Duration)
+
+// StartWithLateFn starts a canary that will call lateFn every expectedDuration
+// period (if not cancelled) and once when cancelled.
+func StartWithLateFn(expectedDuration time.Duration, lateFn CanaryFunc, doneFn CanaryFunc) CancelFunc {
+	if !*enableCanaries {
+		return func() {}
+	}
+
+	done := make(chan struct{})
+	cancel := func() {
+		close(done)
```

Comment:
can you guard against cancel being called multiple times?

Reply:
Done.